### PR TITLE
move locking from context into the the db

### PR DESCRIPTION
**Pull request by michaelmattig**

### CODE REVIEW

**Comment by michaelmattig** [comment]
[DB locking aus dem Kontext in die DBs schieben](https://trello.com/c/9BkwieeL/763-db-locking-aus-dem-kontext-in-die-dbs-schieben)

File: services/src/pro/datasets/in_memory.rs [hunk 1]
```diff
@@ -266,15 +350,21 @@ impl DatasetProvider<UserSession> for ProHashMapDatasetDb {
 
     async fn load(&self, session: &UserSession, dataset: &DatasetId) -> Result<Dataset> {
         ensure!(
-            self.dataset_permissions
+            self.backend
+                .read()
+                .await
```
Comment (ChristianBeilschmidt) [inline services/src/pro/datasets/in_memory.rs:355]: I think you should only get the lock once.

File: services/src/pro/datasets/in_memory.rs [hunk 2]
```diff
@@ -288,15 +378,21 @@ impl DatasetProvider<UserSession> for ProHashMapDatasetDb {
         match dataset {
             DatasetId::Internal { dataset_id: _ } => {
                 ensure!(
-                    self.dataset_permissions
+                    self.backend
+                        .read()
+                        .await
+                        .dataset_permissions
```
Comment (ChristianBeilschmidt) [inline services/src/pro/datasets/in_memory.rs:384]: Same here

File: services/src/pro/datasets/in_memory.rs [hunk 3]
```diff
@@ -372,7 +471,10 @@ impl
         >,
     > {
         ensure!(
-            self.dataset_permissions
+            self.backend
+                .read()
+                .await
+                .dataset_permissions
```
Comment (ChristianBeilschmidt) [inline services/src/pro/datasets/in_memory.rs:477]: and here

File: services/src/pro/datasets/in_memory.rs [hunk 4]
```diff
@@ -409,7 +514,10 @@ impl
     ) -> Result<Box<dyn MetaData<OgrSourceDataset, VectorResultDescriptor, VectorQueryRectangle>>>
     {
         ensure!(
-            self.dataset_permissions
+            self.backend
+                .read()
+                .await
+                .dataset_permissions
```
Comment (ChristianBeilschmidt) [inline services/src/pro/datasets/in_memory.rs:520]: and here

File: services/src/pro/datasets/in_memory.rs [hunk 5]
```diff
@@ -446,7 +557,10 @@ impl
     ) -> Result<Box<dyn MetaData<GdalLoadingInfo, RasterResultDescriptor, RasterQueryRectangle>>>
     {
         ensure!(
-            self.dataset_permissions
+            self.backend
+                .read()
+                .await
+                .dataset_permissions
```
Comment (ChristianBeilschmidt) [inline services/src/pro/datasets/in_memory.rs:563]: and here

File: services/src/pro/projects/hashmap_projectdb.rs [hunk 6]
```diff
@@ -33,12 +34,17 @@ impl ProjectDb<UserSession> for ProHashMapProjectDb {
             offset,
             limit,
         } = options.user_input;
+
+        let all_projects = self.projects.read().await;
+
         #[allow(clippy::flat_map_option)]
         let mut projects = self
             .permissions
+            .read()
```
Comment (ChristianBeilschmidt) [inline services/src/pro/projects/hashmap_projectdb.rs:43]: Why do you have in this case two locks but for other in memory dbs only one lock for everything?

File: services/src/pro/projects/hashmap_projectdb.rs [hunk 7]
```diff
@@ -63,14 +69,14 @@ impl ProjectDb<UserSession> for ProHashMapProjectDb {
...
+        &self,
         session: &UserSession,
         create: Validated<CreateProject>,
     ) -> Result<ProjectId> {
         let project: Project = Project::from_create_project(create.user_input);
         let id = project.id;
-        self.projects.insert(id, vec![project]);
-        self.permissions.push(UserProjectPermission {
+        self.projects.write().await.insert(id, vec![project]);
+        self.permissions.write().await.push(UserProjectPermission {
```
Comment (ChristianBeilschmidt) [inline services/src/pro/projects/hashmap_projectdb.rs:79]: cf. also here

File: services/src/pro/users/hashmap_userdb.rs [hunk 8]
```diff
@@ -16,29 +16,36 @@ use geoengine_datatypes::util::Identifier;
...
     /// Register a user
-    async fn register(&mut self, user_registration: Validated<UserRegistration>) -> Result<UserId> {
+    async fn register(&self, user_registration: Validated<UserRegistration>) -> Result<UserId> {
         let user_registration = user_registration.user_input;
         ensure!(
-            !self.users.contains_key(&user_registration.email),
+            !self
+                .users
+                .read()
+                .await
```
Comment (ChristianBeilschmidt) [inline services/src/pro/users/hashmap_userdb.rs:32]: If you need the write lock anyway, why require a read lock first and then a write lock?

**Review by ChristianBeilschmidt — CHANGES_REQUESTED** [review]

File: services/src/pro/projects/hashmap_projectdb.rs (hunk 6 above)
Comment (michaelmattig) [reply services/src/pro/projects/hashmap_projectdb.rs:43]: it requires only one more lock but saves introducing a backend struct, so: convenience

File: services/src/pro/projects/hashmap_projectdb.rs (hunk 7 above)
Comment (michaelmattig) [reply services/src/pro/projects/hashmap_projectdb.rs:79]: cf my other comment

File: services/src/pro/datasets/in_memory.rs (hunk 1 above)
Comment (michaelmattig) [reply services/src/pro/datasets/in_memory.rs]: fccf47a2459c22ceb51d2ac1d1b4e0237d13cfd1

File: services/src/pro/datasets/in_memory.rs (hunk 2 above)
Comment (michaelmattig) [reply services/src/pro/datasets/in_memory.rs]: fccf47a2459c22ceb51d2ac1d1b4e0237d13cfd1

File: services/src/pro/datasets/in_memory.rs (hunk 3 above)
Comment (michaelmattig) [reply services/src/pro/datasets/in_memory.rs]: fccf47a2459c22ceb51d2ac1d1b4e0237d13cfd1

File: services/src/pro/datasets/in_memory.rs (hunk 4 above)
Comment (michaelmattig) [reply services/src/pro/datasets/in_memory.rs]: fccf47a2459c22ceb51d2ac1d1b4e0237d13cfd1

File: services/src/pro/datasets/in_memory.rs (hunk 5 above)
Comment (michaelmattig) [reply services/src/pro/datasets/in_memory.rs]: fccf47a2459c22ceb51d2ac1d1b4e0237d13cfd1

File: services/src/pro/users/hashmap_userdb.rs (hunk 8 above)
Comment (michaelmattig) [reply services/src/pro/users/hashmap_userdb.rs]: fccf47a2459c22ceb51d2ac1d1b4e0237d13cfd1

**Review by ChristianBeilschmidt — APPROVED** [review]

**Comment by ChristianBeilschmidt** [comment]
bors merge

**Pull request merged**